src/common/lazy_alloc_array.h

```diff
@@ -36,53 +38,104 @@ class LazyAllocArray {
   inline void Clear();
 
+  void SignalForKill();
+
  private:
+  template<typename SyncObject>
+  class unique_unlock {
+   public:
+    explicit unique_unlock(std::unique_lock<SyncObject> *lock)
+      : lock_(lock) {
+      if (lock_) {
+        lock_->unlock();
+      }
+    }
+    ~unique_unlock() {
+      if (lock_) {
+        lock_->lock();
+      }
+    }
+   private:
+    std::unique_lock<SyncObject> *lock_;
+  };
+
   /*! \brief the initial size of the array */
   static constexpr std::size_t kInitSize = 16;
   /*! \brief mutex used during creation */
   std::mutex create_mutex_;
   /*! \brief internal data fir initial size */
-  std::array<std::unique_ptr<TElem>, kInitSize> head_;
+  std::array<std::shared_ptr<TElem>, kInitSize> head_;
   /*! \brief overflow array of more elements */
-  std::vector<std::unique_ptr<TElem> > more_;
+  std::vector<std::shared_ptr<TElem> > more_;
+  /*! \brief Signal shutdown of array */
+  std::atomic<bool> exit_now_;
 };
 
+template<typename TElem>
+inline LazyAllocArray<TElem>::LazyAllocArray()
+  : exit_now_(false) {
+}
+
 // implementations
 template<typename TElem>
 template<typename FCreate>
-inline TElem* LazyAllocArray<TElem>::Get(int index, FCreate creator) {
+inline std::shared_ptr<TElem> LazyAllocArray<TElem>::Get(int index, FCreate creator) {
   CHECK_GE(index, 0);
   size_t idx = static_cast<size_t>(index);
   if (idx < kInitSize) {
-    TElem *ptr = head_[idx].get();
-    if (ptr != nullptr) {
+    std::shared_ptr<TElem> ptr = head_[idx];
+    if (ptr) {
       return ptr;
     } else {
       std::lock_guard<std::mutex> lock(create_mutex_);
-      TElem *ptr = head_[idx].get();
-      if (ptr != nullptr) return ptr;
-      head_[idx].reset(ptr = creator());
-      return ptr;
+      if (!exit_now_.load()) {
+        std::shared_ptr<TElem> ptr = head_[idx];
+        if (ptr) {
+          return ptr;
+        }
+        ptr = head_[idx] = std::shared_ptr<TElem>(creator());
+        return ptr;
+      }
     }
   } else {
     std::lock_guard<std::mutex> lock(create_mutex_);
-    idx -= kInitSize;
-    if (more_.size() <= idx) more_.resize(idx + 1);
-    TElem *ptr = more_[idx].get();
-    if (ptr != nullptr) return ptr;
-    more_[idx].reset(ptr = creator());
-    return ptr;
+    if (!exit_now_.load()) {
+      idx -= kInitSize;
+      if (more_.size() <= idx) {
+        more_.reserve(idx + 1);
+        while (more_.size() <= idx) {
+          more_.push_back(std::shared_ptr<TElem>(nullptr));
+        }
+      }
+      std::shared_ptr<TElem> ptr = more_[idx];
+      if (ptr) {
+        return ptr;
+      }
+      ptr = more_[idx] = std::shared_ptr<TElem>(creator());
+      return ptr;
+    }
   }
+  return nullptr;
 }
 
 template<typename TElem>
 inline void LazyAllocArray<TElem>::Clear() {
-  std::lock_guard<std::mutex> lock(create_mutex_);
+  std::unique_lock<std::mutex> lock(create_mutex_);
+  exit_now_.store(true);
+  // Currently, head_ and more_ never get smaller, so it's safe to
```

Comment:
It's a correct assumption at this time.  There's no existing code which reduces the size of head_ or more_.  But it's actually still ok because we call size() again each pass.  See below for size cacheing.